hadoop-tools/hadoop-azure/src/main/java/org/apache/hadoop/fs/azurebfs/services/TimeoutOptimizer.java

```diff
@@ -0,0 +1,244 @@
+
+    /**
+     * Constructor to initialize the parameters in class,
+     * depending upon what is configured in the settings.
+     * @param url request URL
+     * @param opType operation type
+     * @param retryPolicy retry policy set for this instance of AbfsClient
+     * @param abfsConfiguration current configuration
+     */
+    public TimeoutOptimizer(URL url, AbfsRestOperationType opType, ExponentialRetryPolicy retryPolicy, AbfsConfiguration abfsConfiguration) {
+        this.url = url;
+        this.opType = opType;
+        if (opType != null) {
+            this.retryPolicy = retryPolicy;
+            this.abfsConfiguration = abfsConfiguration;
+            String shouldOptimize = abfsConfiguration.get(ConfigurationKeys.AZURE_OPTIMIZE_TIMEOUTS);
+            if (shouldOptimize == null || shouldOptimize.isEmpty()) {
+                // config is not set
+                this.shouldOptimizeTimeout = false;
+            }
+            else {
+                this.shouldOptimizeTimeout = Boolean.parseBoolean(shouldOptimize);
+                if (this.shouldOptimizeTimeout) {
+                    // config is set to true
+                    if (abfsConfiguration.get(ConfigurationKeys.AZURE_MAX_REQUEST_TIMEOUT) != null) {
+                        this.maxReqTimeout = Integer.parseInt(abfsConfiguration.get(ConfigurationKeys.AZURE_MAX_REQUEST_TIMEOUT));
+                    }
+                    if (abfsConfiguration.get(ConfigurationKeys.AZURE_REQUEST_TIMEOUT_INCREASE_RATE) != null) {
+                        this.timeoutIncRate = Integer.parseInt(abfsConfiguration.get(ConfigurationKeys.AZURE_REQUEST_TIMEOUT_INCREASE_RATE));
+                    }
+                    if (this.maxReqTimeout == -1 || this.timeoutIncRate == -1) {
+                        this.shouldOptimizeTimeout = false;
+                    } else {
+                        initTimeouts();
+                        updateUrl();
+                    }
+                }
+            }
+        } else {
+            // optimization not required for opType == null
+            this.shouldOptimizeTimeout = false;
+        }
+    }
+
+    public void updateRetryTimeout(int retryCount) {
+        if (!this.shouldOptimizeTimeout) {
+            return;
+        }
+
+        // update all timeout values
+        updateTimeouts(retryCount);
+        updateUrl();
+    }
+
+    public URL getUrl() {
+        return url;
+    }
+    public boolean getShouldOptimizeTimeout() {
+        return this.shouldOptimizeTimeout;
+    }
+
+    public int getRequestTimeout() {
+        return requestTimeout;
+    }
+
+    public int getReadTimeout() {
+        return readTimeout;
+    }
+
+    public int getReadTimeout(final int defaultTimeout) {
+        if (readTimeout != -1 && shouldOptimizeTimeout) {
+            return readTimeout;
+        }
+        return defaultTimeout;
+    }
+
+    private void initTimeouts() {
+        if (!shouldOptimizeTimeout) {
```

Comment:
initTimeouts can happen only if shouldOptimizeTimeout is true. Lets remove this if-block?

Reply:
Updated!